language/resource-viewer/src/lib.rs

```diff
@@ -0,0 +1,236 @@
+            annotated_fields.push(self.annotate_value(v, ty)?);
+        }
+        Ok(AnnotatedMoveStruct {
+            is_resource: ty.is_resource,
+            type_: struct_tag,
+            value: field_names
+                .into_iter()
+                .zip(annotated_fields.into_iter())
+                .collect(),
+        })
+    }
+
+    fn annotate_value(&self, value: &MoveValue, ty: &FatType) -> Result<AnnotatedMoveValue> {
+        Ok(match (value, ty) {
+            (MoveValue::Bool(b), FatType::Bool) => AnnotatedMoveValue::Bool(*b),
+            (MoveValue::U8(i), FatType::U8) => AnnotatedMoveValue::U8(*i),
+            (MoveValue::U64(i), FatType::U64) => AnnotatedMoveValue::U64(*i),
+            (MoveValue::U128(i), FatType::U128) => AnnotatedMoveValue::U128(*i),
+            (MoveValue::Address(a), FatType::Address) => AnnotatedMoveValue::Address(*a),
+            (MoveValue::Vector(a), FatType::Vector(ty)) => match ty.as_ref() {
+                FatType::U8 => AnnotatedMoveValue::Bytes(
+                    a.iter()
+                        .map(|v| match v {
+                            MoveValue::U8(i) => Ok(*i),
+                            _ => Err(anyhow!("unexpected value type")),
+                        })
+                        .collect::<Result<_>>()?,
+                ),
+                _ => AnnotatedMoveValue::Vector(
+                    a.iter()
+                        .map(|v| self.annotate_value(v, ty.as_ref()))
+                        .collect::<Result<_>>()?,
+                ),
+            },
+            (MoveValue::Struct(s), FatType::Struct(ty)) => {
+                AnnotatedMoveValue::Struct(self.annotate_struct(s, ty.as_ref())?)
+            }
+            _ => {
+                return Err(anyhow!(
+                    "Cannot annotate value {:?} with type {:?}",
+                    value,
+                    ty
+                ))
+            }
+        })
+    }
+}
+
+fn write_indent(f: &mut Formatter, indent: u64) -> std::fmt::Result {
+    for _i in 0..indent {
+        write!(f, " ")?;
+    }
+    Ok(())
+}
+
+fn pretty_print_value(
+    f: &mut Formatter,
+    value: &AnnotatedMoveValue,
+    indent: u64,
+) -> std::fmt::Result {
+    match value {
+        AnnotatedMoveValue::Bool(b) => write!(f, "{}", b),
+        AnnotatedMoveValue::U8(v) => write!(f, "{}u8", v),
+        AnnotatedMoveValue::U64(v) => write!(f, "{}", v),
+        AnnotatedMoveValue::U128(v) => write!(f, "{}u128", v),
+        AnnotatedMoveValue::Address(a) => write!(f, "0x{}", a),
+        AnnotatedMoveValue::Vector(v) => {
+            writeln!(f, "[")?;
+            for value in v.iter() {
+                write_indent(f, indent + 4)?;
+                pretty_print_value(f, value, indent + 4)?;
+                writeln!(f, ",")?;
+            }
+            write_indent(f, indent)?;
+            write!(f, "]")
+        }
+        AnnotatedMoveValue::Bytes(v) => write!(f, "{}", hex::encode(&v)),
```

Comment:
likewise here

Reply:
I don't think we should trim the leading 0s here IMHO. The length of a byte array is indeed some critical information to things like pubkeys, lcs serialied data, etc.